File: src/scranpy/lib/aggregate_across_cells.cpp
#include "parallel.h" // must be first, to set all macros.

#include "Mattress.h"
#include "scran/aggregation/AggregateAcrossCells.hpp"
#include <cstdint>
#include <algorithm>

//[[export]]
void* combine_factors(int32_t length, int32_t number, const uintptr_t* inputs /** void_p */, int32_t* output_combined /** numpy */) {
    std::vector<const int32_t*> factors;
    factors.reserve(number);
    for (int32_t i = 0; i < number; ++i) {
        factors.push_back(reinterpret_cast<const int32_t*>(inputs[i]));
    }

    auto res = scran::AggregateAcrossCells::combine_factors(length, std::move(factors), output_combined);
    return new decltype(res)(std::move(res));
}

//[[export]]
int32_t get_combined_factors_size(void* ptr) {
    auto combined = reinterpret_cast<scran::AggregateAcrossCells::Combinations<int32_t>*>(ptr);
    return combined->counts.size();
}

//[[export]]
void get_combined_factors_level(void* ptr, int32_t i, int32_t* output /** numpy */) {
    auto combined = reinterpret_cast<const scran::AggregateAcrossCells::Combinations<int32_t>*>(ptr);
    const auto& lev = combined->factors[i];
    std::copy(lev.begin(), lev.end(), output);
}

//[[export]]
void get_combined_factors_count(void* ptr, int32_t* output /** numpy */) {
    auto combined = reinterpret_cast<const scran::AggregateAcrossCells::Combinations<int32_t>*>(ptr);
    std::copy(combined->counts.begin(), combined->counts.end(), output);
}

//[[export]]
void free_combined_factors(void* ptr) {
    delete reinterpret_cast<scran::AggregateAcrossCells::Combinations<int32_t>*>(ptr);
}

//[[export]]
void aggregate_across_cells(
    void* mat,
    const int32_t* groups /** numpy */,
    int32_t ngroups,
    uint8_t do_sums,
    double* output_sums /** void_p */,
    uint8_t do_detected,
    int32_t* output_detected /** void_p */,
    int32_t nthreads)
{
    const auto& ptr = reinterpret_cast<const Mattress*>(mat)->ptr;
    auto NR = ptr->nrow();
    auto NC = ptr->ncol();

    std::vector<double*> output_sums_ptrs;
    if (do_sums) {
        output_sums_ptrs.reserve(ngroups);
        for (int32_t g = 0; g < ngroups; ++g) {
            output_sums_ptrs.push_back(output_sums);
            output_sums += NR;
        }
    }

    std::vector<int32_t*> output_detected_ptrs;
    if (do_detected) {
        output_detected_ptrs.reserve(ngroups);
        for (int32_t g = 0; g < ngroups; ++g) {
            output_detected_ptrs.push_back(output_detected);
            output_detected += NR;
        }
    }

    scran::AggregateAcrossCells aggr;
    aggr.set_num_threads(nthreads);
    aggr.run(ptr.get(), groups, std::move(output_sums_ptrs), std::move(output_detected_ptrs));
    return;
}

File: src/scranpy/aggregation/aggregate_across_cells.py
from numpy import zeros, float64, int32, uintp
from typing import Sequence, Tuple, Union
from dataclasses import dataclass
from biocframe import BiocFrame
from summarizedexperiment import SummarizedExperiment

from .._utils import factorize, tatamize_input, MatrixTypes
from .. import _cpphelpers as lib


class _CombinedFactors:
    def __init__(self, ptr):
        self._ptr = ptr
        self._n = lib.get_combined_factors_size(ptr)

    def __del__(self):
        lib.free_combined_factors(self._ptr)

    def size(self):
        return self._n

    def levels(self, i):
        output = zeros((self._n,), dtype=int32)
        lib.get_combined_factors_level(self._ptr, i, output)
        return output

    def counts(self):
        output = zeros((self._n,), dtype=int32)
        lib.get_combined_factors_count(self._ptr, output)
        return output


@dataclass
class AggregateAcrossCellsOptions:
    """Options to pass to :py:meth:`~scranpy.aggregation.aggregate_across_cells.aggregate_across_cells`.

    Attributes:
        compute_sums:
            Whether to compute the sum of each group.

        compute_detected:
            Whether to compute the number of detected cells in each group.

        assay_type:
            Assay to use from ``input`` if it is a
            :py:class:`~summarizedexperiment.SummarizedExperiment.SummarizedExperiment`.

        num_threads:
            Number of threads.
    """

    compute_sums: bool = True
    compute_detected: bool = True
    assay_type: Union[int, str] = 0
    num_threads: int = 1


def aggregate_across_cells(
    input: MatrixTypes,
    groups: Union[Sequence, Tuple[Sequence], dict, BiocFrame],
    options: AggregateAcrossCellsOptions = AggregateAcrossCellsOptions(),
) -> SummarizedExperiment:
    """Aggregate expression values for groups of cells.

    Args:
        input: Matrix-like object where rows are features and columns are cells, typically containing
            expression values of some kind. This should be a matrix class that can be converted into a
            :py:class:`~mattress.TatamiNumericPointer.TatamiNumericPointer`.

            Alternatively, a :py:class:`~summarizedexperiment.SummarizedExperiment.SummarizedExperiment`
            containing such a matrix in its assays.

            Developers may also provide a :py:class:`~mattress.TatamiNumericPointer.TatamiNumericPointer` directly.

        groups:
            A sequence of length equal to the number of columns of ``input``, specifying the group
            to which each column is assigned. Alternatively, a tuple, dictionary, or
            :py:class:`~biocframe.BiocFrame` of one or more such sequences, in which case
            each unique combination of levels across all sequences is defined as a "group".

        options:
            Further options.

    Returns:
        A SummarizedExperiment where each row corresponds to a row in ``input`` and each
        column corresponds to a group. Assays contain the sum of expression values (if
        ``options.compute_sums = True``) and the number of cells with detected expression (if
        ``options.compute_detected = True``) for each group. Column data contains the identity
        of each group; for ``groups`` containing multiple sequences, the identity of each
        group is defined as a unique combination of levels from each sequence.
    """
    x = tatamize_input(input, options.assay_type)
    NR = x.nrow()
    NC = x.ncol()

    factors = []
    factor_names = None
    if isinstance(groups, tuple):
        factors = [*groups]
    elif isinstance(groups, BiocFrame):
        factor_names = groups.column_names
        for gr in factor_names:
            factors.append(groups.column(gr))
    elif isinstance(groups, dict):
        factor_names = list(groups.keys())
        for gr in factor_names:
            factors.append(groups[gr])
    else:
        factors.append(groups)

    stored_levels = []
    stored_indices = []
    for i, si in enumerate(factors):
        if len(si) != NC:
            raise ValueError(
                "length of grouping vectors should be equal to the number of columns of 'input'"
            )
        lev, ind = factorize(si)
        stored_levels.append(lev)
        stored_indices.append(ind)

    # Possibly combining factors here.
    nstored = len(stored_indices)
    combined = None
    levels = []
    counts = None

    if nstored == 1:
        levels = [stored_levels[0]]
        combined = stored_indices[0]
        counts = zeros((len(levels[0]),), dtype=int32)
        for il in combined:
            counts[il] += 1
    else:
        indptrs = zeros((nstored,), dtype=uintp)
        for i, si in enumerate(stored_indices):
            indptrs[i] = si.ctypes.data

        combined = zeros((NC,), dtype=int32)
        ptr = _CombinedFactors(
            lib.combine_factors(NC, nstored, indptrs.ctypes.data, combined)
        )
        for i, sl in enumerate(stored_levels):
            outlev = ptr.levels(i)
            levels.append([sl[j] for j in outlev])
        counts = ptr.counts()

    ngroups = len(counts)

    # Actually doing the aggregation.
    sums_out = None
    sums_out_ptr = 0
    if options.compute_sums:
        sums_out = zeros((ngroups, NR), dtype=float64)
        sums_out_ptr = sums_out.ctypes.data

    detected_out = None
    detected_out_ptr = 0
    if options.compute_detected:
        detected_out = zeros((ngroups, NR), dtype=int32)
        detected_out_ptr = detected_out.ctypes.data

    lib.aggregate_across_cells(
        x.ptr,
        groups=combined,
        ngroups=ngroups,
        do_sums=options.compute_sums,
        output_sums=sums_out_ptr,
        do_detected=options.compute_detected,
        output_detected=detected_out_ptr,
        nthreads=options.num_threads,
    )

    # Formatting the output.
    output_assays = {}
    if options.compute_sums:
        output_assays["sums"] = sums_out.T
    if options.compute_detected:
        output_assays["detected"] = detected_out.T

    output = SummarizedExperiment(output_assays)
    if nstored == 1:
        output.column_names = levels[0]

    if factor_names is None:
        factor_names = ["factor_" + str(i + 1) for i in range(nstored)]

    reported_factors = {}
    for i, x in enumerate(levels):
        reported_factors[factor_names[i]] = x
    reported_factors["counts"] = counts
    output.column_data = BiocFrame(reported_factors)

    return output
